cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/MapperTransformExecutorFactory.java

```diff
@@ -0,0 +1,133 @@
+  }
+
+  @Override
+  protected BatchRuntimeContext createRuntimeContext(String stageName) {
+    Map<String, String> stageRuntimeArgs = pluginRuntimeArgs.get(stageName);
+    if (stageRuntimeArgs == null) {
+      stageRuntimeArgs = new HashMap<>();
+    }
+    return new MapReduceRuntimeContext(taskContext, metrics, new DatasetContextLookupProvider(taskContext),
+                                       stageName, stageRuntimeArgs);
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  protected Transformation getTransformation(String pluginType, String stageName) throws Exception {
+    if (BatchAggregator.PLUGIN_TYPE.equals(pluginType)) {
+      BatchAggregator<?, ?, ?> batchAggregator = pluginInstantiator.newPluginInstance(stageName);
+      BatchRuntimeContext runtimeContext = createRuntimeContext(stageName);
+      batchAggregator.initialize(runtimeContext);
+      return new AggregatorTransformation(batchAggregator,
+                                          new DefaultStageMetrics(metrics, stageName),
+                                          mapOutputKeyClassName,
+                                          mapOutputValClassName);
+    }
+    return super.getTransformation(pluginType, stageName);
+  }
+
+  /**
+   * A Transformation that uses an aggregator's groupBy method. Supports applying a function to the types
+   * returned by the aggregator. These functions are used when the aggregator outputs group keys that are not
+   * WritableComparable and values that are not Writable. For example, aggregators that output StructuredRecord
+   * will need some function to change a StructuredRecord to a StructuredRecordWritable so that we can use this
+   * in mapreduce.
+   *
+   * @param <GROUP_KEY> type of group key output by the aggregator
+   * @param <GROUP_VAL> type of group value used by the aggregator
+   * @param <OUT_KEY> type of output key for mapreduce. Must implement WritableComparable
+   * @param <OUT_VAL> type of output value for mapreduce. Must implement Writable
+   */
+  private static class AggregatorTransformation<GROUP_KEY, GROUP_VAL, OUT_KEY extends Writable,
+    OUT_VAL extends Writable> implements Transformation<GROUP_VAL, KeyValue<OUT_KEY, OUT_VAL>> {
+    private final Aggregator<GROUP_KEY, GROUP_VAL, ?> aggregator;
+    private final StageMetrics stageMetrics;
+    private final WritableConversion<GROUP_KEY, OUT_KEY> keyConversion;
+    private final WritableConversion<GROUP_VAL, OUT_VAL> valConversion;
+
+    public AggregatorTransformation(Aggregator<GROUP_KEY, GROUP_VAL, ?> aggregator,
+                                    StageMetrics stageMetrics,
+                                    String groupKeyClassName,
+                                    String groupValClassName) {
+      this.aggregator = aggregator;
+      this.stageMetrics = stageMetrics;
+      WritableConversion<GROUP_KEY, OUT_KEY> keyConversion = WritableConversions.getConversion(groupKeyClassName);
```

Comment:
can this return null? It does here: https://github.com/caskdata/cdap/pull/5341/files#diff-f7cf8103359905eb8aeacfc3074876b0R244